File: android/src/main/java/dev/novals/capupdate/BundleManager.java
package dev.novals.capupdate;

import android.content.Context;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

/**
 * Manages downloading, extracting, and tracking OTA bundles on the filesystem.
 * Uses only built-in Java ZIP APIs — no external dependencies.
 */
public class BundleManager {

    private static final String BUNDLES_DIR = "cap_update_bundles";
    private final Context context;

    public BundleManager(Context context) {
        this.context = context;
    }

    /**
     * Get the root directory where all bundles are stored.
     */
    public File getBundlesDir() {
        File dir = new File(context.getFilesDir(), BUNDLES_DIR);
        if (!dir.exists()) {
            dir.mkdirs();
        }
        return dir;
    }

    /**
     * Download a ZIP from the given URL, extract it, and return the resolved bundle ID.
     *
     * @param urlString  The download URL.
     * @param bundleId   Optional bundle ID. If null, derived from URL filename.
     * @param checksum   Optional SHA-256 checksum for integrity verification.
     * @return The resolved bundle ID.
     */
    public String downloadAndExtract(String urlString, String bundleId, String checksum) throws Exception {
        URL url = new URL(urlString);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setConnectTimeout(30000);
        connection.setReadTimeout(60000);
        connection.connect();

        int responseCode = connection.getResponseCode();
        if (responseCode != HttpURLConnection.HTTP_OK) {
            throw new Exception("Server returned HTTP " + responseCode + " " + connection.getResponseMessage());
        }

        // Resolve bundle ID
        if (bundleId == null || bundleId.isEmpty()) {
            // Try to get from response header first
            String headerBundleId = connection.getHeaderField("X-Bundle-Id");
            if (headerBundleId != null && !headerBundleId.isEmpty()) {
                bundleId = headerBundleId;
            } else {
                bundleId = deriveBundleIdFromUrl(urlString);
            }
        }

        // Download to temp file
        File tempZip = new File(context.getCacheDir(), "cap_update_" + bundleId + ".zip");
        try (InputStream input = new BufferedInputStream(connection.getInputStream());
             FileOutputStream output = new FileOutputStream(tempZip)) {
            byte[] buffer = new byte[8192];
            int count;
            while ((count = input.read(buffer)) != -1) {
                output.write(buffer, 0, count);
            }
        }

        // Verify checksum if provided
        if (checksum != null && !checksum.isEmpty()) {
            String computed = computeSha256(tempZip);
            if (!computed.equalsIgnoreCase(checksum)) {
                tempZip.delete();
                throw new Exception("Checksum mismatch. Expected: " + checksum + ", got: " + computed);
            }
        }

        // Prepare target directory (overwrite if exists)
        File targetDir = new File(getBundlesDir(), bundleId);
        if (targetDir.exists()) {
            deleteRecursive(targetDir);
        }
        targetDir.mkdirs();

        // Extract
        unzip(tempZip, targetDir);
        tempZip.delete();

        return bundleId;
    }

    /**
     * Get the absolute path to a bundle's directory, or null if it doesn't exist.
     */
    public String getBundlePath(String bundleId) {
        File dir = new File(getBundlesDir(), bundleId);
        return dir.exists() ? dir.getAbsolutePath() : null;
    }

    /**
     * List all downloaded bundle IDs.
     */
    public List<String> getBundleList() {
        List<String> list = new ArrayList<>();
        File[] files = getBundlesDir().listFiles();
        if (files != null) {
            for (File f : files) {
                if (f.isDirectory()) {
                    list.add(f.getName());
                }
            }
        }
        return list;
    }

    /**
     * Delete a bundle's directory.
     */
    public void deleteBundle(String bundleId) {
        File dir = new File(getBundlesDir(), bundleId);
        if (dir.exists()) {
            deleteRecursive(dir);
        }
    }

    /**
     * Recursively search for the directory containing index.html.
     * Handles cases where the ZIP has a wrapper directory.
     */
    public File findWebRoot(File dir) {
        if (new File(dir, "index.html").exists()) {
            return dir;
        }
        File[] children = dir.listFiles();
        if (children != null) {
            for (File child : children) {
                if (child.isDirectory()) {
                    File found = findWebRoot(child);
                    if (found != null) {
                        return found;
                    }
                }
            }
        }
        return null;
    }

    // ──────────────────────────────────────────────
    // Private Helpers
    // ──────────────────────────────────────────────

    private String deriveBundleIdFromUrl(String url) {
        String filename = url.substring(url.lastIndexOf('/') + 1);
        // Strip query string
        int queryIndex = filename.indexOf('?');
        if (queryIndex > 0) {
            filename = filename.substring(0, queryIndex);
        }
        // Strip .zip extension
        if (filename.toLowerCase().endsWith(".zip")) {
            filename = filename.substring(0, filename.length() - 4);
        }
        // Sanitize
        return filename.replaceAll("[^a-zA-Z0-9._-]", "_");
    }

    private void unzip(File zipFile, File targetDir) throws IOException {
        try (ZipInputStream zis = new ZipInputStream(new BufferedInputStream(new FileInputStream(zipFile)))) {
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                File file = new File(targetDir, entry.getName());

                // Security: prevent zip slip attack
                String canonicalTarget = targetDir.getCanonicalPath();
                String canonicalFile = file.getCanonicalPath();
                if (!canonicalFile.startsWith(canonicalTarget + File.separator) && !canonicalFile.equals(canonicalTarget)) {
                    throw new IOException("Zip entry outside target directory: " + entry.getName());
                }

                if (entry.isDirectory()) {
                    file.mkdirs();
                } else {
                    // Ensure parent dirs exist
                    File parent = file.getParentFile();
                    if (parent != null && !parent.exists()) {
                        parent.mkdirs();
                    }
                    try (FileOutputStream fos = new FileOutputStream(file)) {
                        byte[] buffer = new byte[8192];
                        int count;
                        while ((count = zis.read(buffer)) != -1) {
                            fos.write(buffer, 0, count);
                        }
                    }
                }
            }
        }
    }

    private String computeSha256(File file) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        try (InputStream fis = new BufferedInputStream(new FileInputStream(file))) {
            byte[] buffer = new byte[8192];
            int count;
            while ((count = fis.read(buffer)) != -1) {
                digest.update(buffer, 0, count);
            }
        }
        byte[] hash = digest.digest();
        StringBuilder sb = new StringBuilder();
        for (byte b : hash) {
            sb.append(Integer.toString((b & 0xff) + 0x100, 16).substring(1));
        }
        return sb.toString();
    }

    private void deleteRecursive(File fileOrDir) {
        if (fileOrDir.isDirectory()) {
            File[] children = fileOrDir.listFiles();
            if (children != null) {
                for (File child : children) {
                    deleteRecursive(child);
                }
            }
        }
        fileOrDir.delete();
    }
}

File: ios/Sources/CapUpdatePlugin/BundleManager.swift
import CommonCrypto
import Foundation
import SSZipArchive

/// Manages downloading, extracting, and tracking OTA bundles on the filesystem.
public class BundleManager {

    public static let shared = BundleManager()

    private static let bundlesDirName = "cap_update_bundles"

    public init() {}

    /// Root directory where all bundles are stored.
    public func getBundlesDir() -> URL {
        let filesDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        let dir = filesDir.appendingPathComponent(Self.bundlesDirName, isDirectory: true)
        if !FileManager.default.fileExists(atPath: dir.path) {
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        return dir
    }

    /// Download a ZIP from the given URL, extract it, and return the resolved bundle ID.
    public func downloadAndExtract(urlString: String, bundleId: String?, checksum: String?) throws -> String {
        guard let url = URL(string: urlString) else {
            throw BundleError.invalidUrl
        }

        // Synchronous download (called from background queue)
        let (data, response, error) = URLSession.shared.synchronousDataTask(with: url)

        if let error = error {
            throw BundleError.downloadFailed(error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            throw BundleError.downloadFailed("HTTP \(code)")
        }

        guard let data = data else {
            throw BundleError.downloadFailed("No data received")
        }

        // Resolve bundle ID
        var resolvedId = bundleId ?? ""
        if resolvedId.isEmpty {
            // Try response header, then derive from URL
            if let headerBundleId = (response as? HTTPURLResponse)?.value(forHTTPHeaderField: "X-Bundle-Id"),
               !headerBundleId.isEmpty {
                resolvedId = headerBundleId
            } else {
                resolvedId = deriveBundleId(from: urlString)
            }
        }

        // Write to temp file
        let tempDir = FileManager.default.temporaryDirectory
        let tempZip = tempDir.appendingPathComponent("cap_update_\(resolvedId).zip")
        try data.write(to: tempZip)

        // Verify checksum
        if let checksum = checksum, !checksum.isEmpty {
            let computed = try computeSha256(file: tempZip)
            if computed.lowercased() != checksum.lowercased() {
                try? FileManager.default.removeItem(at: tempZip)
                throw BundleError.checksumMismatch(expected: checksum, actual: computed)
            }
        }

        // Prepare target directory
        let targetDir = getBundlesDir().appendingPathComponent(resolvedId, isDirectory: true)
        if FileManager.default.fileExists(atPath: targetDir.path) {
            try FileManager.default.removeItem(at: targetDir)
        }
        try FileManager.default.createDirectory(at: targetDir, withIntermediateDirectories: true)

        // Extract using SSZipArchive
        let success = SSZipArchive.unzipFile(atPath: tempZip.path, toDestination: targetDir.path)
        try? FileManager.default.removeItem(at: tempZip)

        if !success {
            try? FileManager.default.removeItem(at: targetDir)
            throw BundleError.extractionFailed
        }

        return resolvedId
    }

    /// Get the absolute path to a bundle's directory, or nil if it doesn't exist.
    public func getBundlePath(bundleId: String) -> String? {
        let dir = getBundlesDir().appendingPathComponent(bundleId, isDirectory: true)
        return FileManager.default.fileExists(atPath: dir.path) ? dir.path : nil
    }

    /// List all downloaded bundle IDs.
    public func getBundleList() -> [String] {
        let dir = getBundlesDir()
        guard let contents = try? FileManager.default.contentsOfDirectory(
            at: dir,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles]
        ) else {
            return []
        }

        return contents.compactMap { url in
            var isDir: ObjCBool = false
            if FileManager.default.fileExists(atPath: url.path, isDirectory: &isDir), isDir.boolValue {
                return url.lastPathComponent
            }
            return nil
        }
    }

    /// Delete a bundle directory.
    public func deleteBundle(bundleId: String) {
        let dir = getBundlesDir().appendingPathComponent(bundleId, isDirectory: true)
        try? FileManager.default.removeItem(at: dir)
    }

    /// Recursively search for the directory containing index.html.
    public func findWebRoot(dir: URL) -> URL? {
        let indexUrl = dir.appendingPathComponent("index.html")
        if FileManager.default.fileExists(atPath: indexUrl.path) {
            return dir
        }

        guard let contents = try? FileManager.default.contentsOfDirectory(
            at: dir,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles]
        ) else {
            return nil
        }

        for url in contents {
            var isDir: ObjCBool = false
            if FileManager.default.fileExists(atPath: url.path, isDirectory: &isDir), isDir.boolValue {
                if let found = findWebRoot(dir: url) {
                    return found
                }
            }
        }

        return nil
    }

    // MARK: - Private Helpers

    private func deriveBundleId(from urlString: String) -> String {
        var filename = (urlString as NSString).lastPathComponent

        // Strip query string
        if let queryIndex = filename.firstIndex(of: "?") {
            filename = String(filename[..<queryIndex])
        }

        // Strip .zip extension
        if filename.lowercased().hasSuffix(".zip") {
            filename = String(filename.dropLast(4))
        }

        // Sanitize
        let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "._-"))
        return filename.unicodeScalars.map { allowed.contains($0) ? String($0) : "_" }.joined()
    }

    private func computeSha256(file: URL) throws -> String {
        let data = try Data(contentsOf: file)
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes { buffer in
            _ = CC_SHA256(buffer.baseAddress, CC_LONG(data.count), &hash)
        }
        return hash.map { String(format: "%02x", $0) }.joined()
    }
}

// MARK: - Error Types

enum BundleError: LocalizedError {
    case invalidUrl
    case downloadFailed(String)
    case checksumMismatch(expected: String, actual: String)
    case extractionFailed

    var errorDescription: String? {
        switch self {
        case .invalidUrl:
            return "Invalid download URL"
        case .downloadFailed(let reason):
            return "Download failed: \(reason)"
        case .checksumMismatch(let expected, let actual):
            return "Checksum mismatch. Expected: \(expected), got: \(actual)"
        case .extractionFailed:
            return "Failed to extract ZIP archive"
        }
    }
}

// MARK: - URLSession Synchronous Extension

extension URLSession {
    func synchronousDataTask(with url: URL) -> (Data?, URLResponse?, Error?) {
        var data: Data?
        var response: URLResponse?
        var error: Error?

        let semaphore = DispatchSemaphore(value: 0)
        let task = self.dataTask(with: url) { d, r, e in
            data = d
            response = r
            error = e
            semaphore.signal()
        }
        task.resume()
        semaphore.wait()

        return (data, response, error)
    }
}
